### ISSUE TITLE
"Reduce" node and reacting to silent periods

### ISSUE THREAD

**Issue by sputnik13** [influxdata/kapacitor#1304]
We're wanting to do something like the following...

```
stream
  | from()
      .measurement('some_measurement')
  | where(lambda: "some condition")
  | stats(1s)
  | derivative('emitted')
      .unit(1s)
      .nonNegative()
  | window()
      .periodCount('3600s')
      .everyCount(1)
  | reduce(lambda: \"emitted\" + \"_previous\")        <-- new node type
      .as("count")
  | alert()
      .crit(lambda: \"count\" > 0)
```

here "reduce" would take a batch, apply the lambda against each point, and supply intermediate values to the lambda as "_previous" (open to suggestion for a better name)

As for why we want this, we're trying to apply rules/tasks that react to messages/events/metrics that come and go (max-throughput, deadman, etc), we're having to deal with silent periods where the data we're trying to alert on doesn't just go down in frequency but goes also doesn't emit at all for extended periods.

We also have requirements to alert in a timely manner (e.g. evaluate max-throughput over a 1 hour window but alert within a few seconds of alert being triggered) that forces us to use a sliding window rather than stats based approach to evaluating alert conditions.

After some thought along the lines of joining the evaluated stream with a separate stats stream to force evaluation of the alert logic, we discarded this approach as it requires lambda expressions to be modified to account for the prefixes applied by the join node to each data field.  Also creating a stream just to join it and discard it to force evaluation of the pipeline just felt...  dirty.

We're willing to do the implementation of the node, we're looking for feedback as to whether this seems useful enough to include in kapacitor proper.

**Comment by sputnik13** [comment]
looks like we can use sum('emitted').as('count') instead...  I guess I shouldn't be making decisions at 2 am :)

although, a general reduce function does seem useful, it's not absolutely necessary for our current needs, so...  closing request

**Comment by sputnik13** [comment]
So we're in need of this functionality again, due to some changing requirements...  I think this could be either a separate reduce node or a modified eval node.  Having a separate reduce node would probably duplicate some code from the eval node but from the user's standpoint, the symantics would be simpler.

**Comment by nathanielc** [comment]
@sputnik13 I want to make sure I understand the request. Basically the reduce node will maintain a single point and reduce each new point into that point via the provided lambda expression?

Meaning I could implement a `count` function by doing something like this?

```
|reduce(lambda: "_previous" + 1)
```

And a sum could be (assuming the data has a field named `value`):

```
|reduce(lambda: "_previous" + "value")
```

Am I on the right page?  If so what kinds of reduce operations are looking to do that are not already provided with the builtin in nodes like `count` and `sum`?

**Comment by sputnik13** [comment]
@nathanielc yes it's along those lines, the specific thing we need is actually more like the following

```
|reduce(lambda: "_previous.count" + 1, lambda: "originalMessage")
```

I took a stab at it, please take a look when you have a moment
